o1vm/src/mips/constraints.rs

```diff
@@ -482,67 +493,66 @@ impl<Fp: Field> InterpreterEnv for Env<Fp> {
+                        * (num_preimage_bytes_read.clone() - Expr::from(1))
+                        * (num_preimage_bytes_read.clone() - Expr::from(2))
+                        * (num_preimage_bytes_read.clone() - Expr::from(3))
+                        * (num_preimage_bytes_read.clone() - Expr::from(4)),
                 );
 
                 // When at least has_2_byte, then any number of bytes can be
-                // read except 1
-                self.constraints.push(
+                // read from the preimage except 1
+                self.add_constraint(
                     has_n_bytes[1].clone()
-                        * (num_read_bytes.clone() - Expr::from(2))
-                        * (num_read_bytes.clone() - Expr::from(3))
-                        * (num_read_bytes.clone() - Expr::from(4)),
+                        * (num_preimage_bytes_read.clone() - Expr::from(2))
+                        * (num_preimage_bytes_read.clone() - Expr::from(3))
+                        * (num_preimage_bytes_read.clone() - Expr::from(4)),
                 );
                 // When at least has_3_byte, then any number of bytes can be
-                // read except 1 nor 2
-                self.constraints.push(
+                // read from the preimage except 1 nor 2
+                self.add_constraint(
                     has_n_bytes[2].clone()
-                        * (num_read_bytes.clone() - Expr::from(3))
-                        * (num_read_bytes.clone() - Expr::from(4)),
+                        * (num_preimage_bytes_read.clone() - Expr::from(3))
+                        * (num_preimage_bytes_read.clone() - Expr::from(4)),
                 );
 
-                // When has_4_byte, then only can read 4
-                self.constraints
-                    .push(has_n_bytes[3].clone() * (num_read_bytes.clone() - Expr::from(4)));
+                // When has_4_byte, then only can read 4 preimage bytes
+                self.add_constraint(
+                    has_n_bytes[3].clone() * (num_preimage_bytes_read.clone() - Expr::from(4)),
+                );
             }
         }
 
+        // FIXED LOOKUPS
+        // Check that 0 <= preimage read <= actual read <= len <= 4
+        self.add_lookup(Lookup::write_one(
+            LookupTableIDs::AtMost4Lookup,
+            vec![len.clone()],
+        ));
+        self.add_lookup(Lookup::write_one(
+            LookupTableIDs::AtMost4Lookup,
+            vec![actual_read_bytes.clone()],
+        ));
+        self.add_lookup(Lookup::write_one(
+            LookupTableIDs::AtMost4Lookup,
+            vec![num_preimage_bytes_read.clone()],
+        ));
+        self.add_lookup(Lookup::write_one(
+            LookupTableIDs::AtMost4Lookup,
+            vec![len.clone() - actual_read_bytes.clone()],
+        ));
+        self.add_lookup(Lookup::write_one(
+            LookupTableIDs::AtMost4Lookup,
+            vec![actual_read_bytes.clone() - num_preimage_bytes_read.clone()],
```

Comment:
Note (for me): I always forget the difference between these two variables. The variable `actual_read_bytes` is the number of bytes that it reads from the oracle, which can be either the first 8 bytes that contain the length or the actual bytes of the preimage. `num_preimage_bytes_read` is only non-zero when we are actually getting values of the preimage, and it will be maximum the number of bytes we actually read from the preimage. I think we still need to verify later the whole semantics. I am quite sure we can simplify the logic.